### PR TITLE
CFINSPEC-13:  Adds support for specifying gem dependencies in the metadata file of InSpec Profile.

### PR DESCRIPTION
**Pull request by Vasu1105**
Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

## Description

This PR adds the support for specifying gem dependencies in the metadata file of Profile.
* Handled version constraint like "=  1.0.0", "~> 1.0.0", ">= 2.0.0, <= 3.0.0" etc.
* Handls gem dependencies for dependent profiles.
* Adds Test Cases
* Added docs.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:

- [ ] I have read the **CONTRIBUTING** document.

### CODE REVIEW

**Comment by aaronlippold** [comment]
I think this is great. 

Did we think through how this will be supported or what warnings we will give to folks if they are using the `--archieve` command. Will we pull and store the gem in the archieve for `offline install` on site?

**Comment by Vasu1105** [comment]
@aaronlippold currently we are only focusing on installing the gems from the RubyGems. We can think about other possibilities of this in the future. If you have more ideas around this please feel free to add the issue with your thoughts we will try to accommodate the same whenever we get the bandwidth.

File: test/fixtures/profiles/profile-with-gem-dependency/controls/example.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+require "money"
```
Comment (clintoncwolfe) [inline test/fixtures/profiles/profile-with-gem-dependency/controls/example.rb:1]: Looks like we need to implement something for `inspec check`, too:
```
[cwolfe@lodi inspec-5]$ bundle exec inspec check test/fixtures/profiles/profile-with-gem-dependency
bundler: failed to load command: inspec (/Users/cwolfe/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/bin/inspec)
LoadError: cannot load such file -- money
  /Users/cwolfe/sandbox/inspec/inspec-5/lib/inspec/dsl_shared.rb:33:in `require'
  /Users/cwolfe/sandbox/inspec/inspec-5/lib/inspec/dsl_shared.rb:33:in `require'
```

File: test/fixtures/profiles/profile-with-gem-dependency/inspec.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+copyright: The Authors
+copyright_email: you@example.com
+license: Apache-2.0
+summary: An InSpec Compliance Profile
+version: 0.1.0
+supports:
+  platform: os
+gem_dependencies:
+- name: money
+  version: "= 6.16.0"
```
Comment (clintoncwolfe) [inline test/fixtures/profiles/profile-with-gem-dependency/inspec.yml]: I think this should be loosened to be ">= 6.16.0" - pinning it "=" may break in the future.

**Review by clintoncwolfe — COMMENTED** [review]
I can't get this to work locally. 
I have 
```
[cwolfe@lodi inspec-5]$ find ~/.inspec/gems -name 'money-*'
/Users/cwolfe/.inspec/gems/2.7.0/specifications/money-6.16.0.gemspec
/Users/cwolfe/.inspec/gems/2.7.0/cache/money-6.16.0.gem
/Users/cwolfe/.inspec/gems/2.7.0/gems/money-6.16.0
```
and
```
[cwolfe@lodi inspec-5]$ bundle exec inspec exec test/fixtures/profiles/profile-with-gem-dependency
Unable to resolve dependency: user requested 'money (= 6.16.0)'
```

File: test/fixtures/profiles/profile-with-gem-dependency/controls/example.rb (hunk 1 above)
Comment (Vasu1105) [reply test/fixtures/profiles/profile-with-gem-dependency/controls/example.rb:1]: Fixed this. Thanks, @clintoncwolfe !!

File: test/fixtures/profiles/profile-with-gem-dependency/inspec.yml (hunk 2 above)
Comment (Vasu1105) [reply test/fixtures/profiles/profile-with-gem-dependency/inspec.yml]: Updated to ">= 6.16.0"

**Comment by Vasu1105** [comment]
> I can't get this to work locally. I have
> 
> ```
> [cwolfe@lodi inspec-5]$ find ~/.inspec/gems -name 'money-*'
> /Users/cwolfe/.inspec/gems/2.7.0/specifications/money-6.16.0.gemspec
> /Users/cwolfe/.inspec/gems/2.7.0/cache/money-6.16.0.gem
> /Users/cwolfe/.inspec/gems/2.7.0/gems/money-6.16.0
> ```
> 
> and
> 
> ```
> [cwolfe@lodi inspec-5]$ bundle exec inspec exec test/fixtures/profiles/profile-with-gem-dependency
> Unable to resolve dependency: user requested 'money (= 6.16.0)'
> ```

Thanks, @clintoncwolfe, this is resolved now. Earlier I was unable to reproduce this on my local and had this issue only on Windows, now able to reproduce on my local too. This is fixed now.

**Comment by clintoncwolfe** [comment]
~Well as of 6db63ff it no longer errors, but it does not execute the profile, either.~ (fixed, my config error)

`inspec check` does work better, but it does not accept `--auto-install-gems`, which it should.

**Comment by Vasu1105** [comment]
> Well as of [6db63ff](https://github.com/inspec/inspec/commit/6db63ffc1a54e89a618828e870031a691c62b362) it no longer errors, but it does not execute the profile, either.
> 
> `inspec check` does work better, but it does not accept `--auto-install-gems`, which it should.
Added the option for `inspec check`. The reason for profile nonexecution is because there is a reporter set in the config.json file of inspec.

**Review by clintoncwolfe — APPROVED** [review]
This is fantastic. We've looked for this feature for a long time.

**Pull request merged**